src/sqlfluff/rules/L042.py

```diff
@@ -48,44 +78,337 @@ class Rule_L042(BaseRule):
+        is_select = segment.all(is_type(*select_types))
+        is_select_child = parent_stack.any(is_type(*select_types))
+        if not is_select or is_select_child:
+            # Subvert the Crawler
+            return None
+
+        # Gather all possible offending Elements in one crawl
+        nested_subqueries: List[_NestedSubQuerySummary] = []
+        selects = segment.recursive_crawl(*select_types, recurse_into=True)
+        for select in selects.iterate_segments():
+            for res in _find_nested_subqueries(select):
+                if res.parent_clause_type not in parent_types:
+                    continue
+                nested_subqueries.append(res)
+
+        if not nested_subqueries:
+            return None
+        # If there are offending elements calculate fixes
+        return _calculate_fixes(
+            dialect=context.dialect,
+            root_select=segment,
+            nested_subqueries=nested_subqueries,
+        )
+
+
+def _calculate_fixes(
+    dialect: Dialect,
+    root_select: Segments,
+    nested_subqueries: List[_NestedSubQuerySummary],
+) -> List[LintResult]:
+    """Given the Root select and the offending subqueries calculate fixes."""
+    is_with = root_select.all(is_type("with_compound_statement"))
+    # TODO: consider if we can fix recursive CTEs
+    is_recursive = is_with and len(root_select.children(is_name("recursive"))) > 0
+    case_preference = _get_case_preference(root_select)
+    # generate an instance which will track and shape out output CTE
+    ctes = _CTEBuilder()
+    # Init the output/final select &
+    # populate existing CTEs
+    for cte in root_select.children(is_type("common_table_expression")):
+        assert isinstance(cte, CTEDefinitionSegment), "TypeGuard"
+        ctes.insert_cte(cte)
+
+    output_select = root_select
+    if is_with:
+        output_select = root_select.children(
+            is_type(
+                "set_expression",
+                "select_statement",
+            )
+        )
+
+    lint_results: List[LintResult] = []
+    mutations_buffer: List[Tuple[BaseSegment, BaseSegment]] = []
+    for parent_type, _, this_seg, subquery in nested_subqueries:
+        alias_name = ctes.create_cte_alias(
+            this_seg.children(is_type("alias_expression"))
+        )
+        new_cte = _create_cte_seg(
+            alias_name=alias_name,
+            subquery=subquery,
+            case_preference=case_preference,
+            dialect=dialect,
+        )
+        ctes.insert_cte(new_cte)
+        # We are preping a mutative change
+        # I gather this is bad. We must provide a position marker
+        # TODO: alternative to mutative change, Dummy pos markers
+        original_pos = this_seg[0].pos_marker
+        assert original_pos, "TypeGuard"
+        mutations_buffer.append(
+            (
+                this_seg[0],
+                _create_table_ref(alias_name, dialect, original_pos),
+            )
+        )
```

Comment:
This Position jihacking Is what was causing the issue with 46 and its guard on position markers.
I'd prefer to alter the from/join tble expression and create the WITH wrapper. However the later means the formed "doesn't" exist and not applying both together makes and invalid parse tree
@barrywhart

Reply:
What do you mean here?
>formed "doesn't" exist

Reply:
Oh, probably "former". 👍

Reply:
@alanmcruickshank: Do you have any thoughts on this? In this PR, he's adding a fix that needs to make larger, multi-level changes to the parse tree, and we're having some new challenges I'm not familiar with.

Reply:
Yea, we need to replace one element within the parse tree & nest the whole tree, within a new object.
Both of these can't be applied at the same time.